google-ads-examples/src/main/java/com/google/ads/googleads/examples/advancedoperations/AddRealEstateFeed.java

```diff
@@ -0,0 +1,507 @@
+            .addAttributes(listingIdAttribute)
+            .addAttributes(listingNameAttribute)
+            .addAttributes(finalUrlsAttribute)
+            .addAttributes(imageUrlAttribute)
+            .addAttributes(contextualKeywordsAttribute)
+            .build();
+
+    // Creates the operation.
+    FeedOperation operation = FeedOperation.newBuilder().setCreate(feed).build();
+
+    // Creates the feed service client.
+    try (FeedServiceClient feedServiceClient =
+        googleAdsClient.getLatestVersion().createFeedServiceClient()) {
+      // Adds the feed.
+      MutateFeedsResponse response =
+          feedServiceClient.mutateFeeds(String.valueOf(customerId), ImmutableList.of(operation));
+      String feedResourceName = response.getResults(0).getResourceName();
+      // Displays the result.
+      System.out.printf("Feed with resource name '%s' was created.%n", feedResourceName);
+      return feedResourceName;
+    }
+  }
+
+  /**
+   * Retrieves details about a feed.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID in which to create criterion.
+   * @param feedResourceName the resource name of the feed.
+   * @return a Map containing the DsaPageFeedCriterionField and FeedAttribute.
+   */
+  private Map<RealEstatePlaceholderField, FeedAttribute> getFeed(
+      GoogleAdsClient googleAdsClient, long customerId, String feedResourceName) {
+    // Constructs the query.
+    String query =
+        "SELECT feed.attributes FROM feed WHERE feed.resource_name = '" + feedResourceName + "'";
+
+    // Constructs the request.
+    SearchGoogleAdsRequest request =
+        SearchGoogleAdsRequest.newBuilder()
+            .setCustomerId(String.valueOf(customerId))
+            .setPageSize(PAGE_SIZE)
+            .setQuery(query)
+            .build();
+
+    // Issues the search request.
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      SearchPagedResponse searchPagedResponse = googleAdsServiceClient.search(request);
+      // Gets the first result because we only need the single feed item we created previously.
+      GoogleAdsRow googleAdsRow = searchPagedResponse.getPage().getResponse().getResults(0);
+      // Gets the attributes list from the feed and creates a map with keys of each attribute and
+      // values of each corresponding ID.
+      List<FeedAttribute> feedAttributeList = googleAdsRow.getFeed().getAttributesList();
+      Map<String, Long> feedAttributeMap = new HashMap<>();
+      for (FeedAttribute feedAttribute : feedAttributeList) {
+        feedAttributeMap.put(feedAttribute.getName().getValue(), feedAttribute.getId().getValue());
+      }
+
+      // Creates a map to return.
+      Map<RealEstatePlaceholderField, FeedAttribute> feedAttributes = new HashMap<>();
+
+      String listingIdName = "Listing ID";
```

Comment:
I'm a bit confused about the manual building of FeedAttribute objects here. Why not use the FeedAttribute objects returned in the query?

Reply:
Excellent question 😄